core/types/block_ext_test.go

```diff
@@ -106,3 +111,410 @@ func assertDifferentPointers[T any](t *testing.T, a *T, b any) {
+func TestBlockBody(t *testing.T) {
+	t.Parallel()
+
+	header := &Header{ParentHash: common.Hash{1}}
+	tx := NewTransaction(0, common.Address{1}, big.NewInt(2), 3, big.NewInt(4), []byte{5})
+	txs := []*Transaction{tx}
+	uncles := []*Header{{ParentHash: common.Hash{6}}}
+	receipts := []*Receipt{{PostState: []byte{7}}}
+	block := NewBlock(header, txs, uncles, receipts, stubHasher{})
+	blockExtras := &BlockBodyExtra{
+		Version: 8,
+		ExtData: ptrTo([]byte{9}),
+	}
+	allExportedFieldsSet(t, blockExtras) // make sure each field is checked
+	extras.Block.Set(block, blockExtras)
+
+	wantUncle := &Header{
+		ParentHash: common.Hash{6},
+		Difficulty: new(big.Int),
+		Number:     new(big.Int),
+	}
+	// uncle returned from Body() has its extra set, even if the block uncle did not have it set
+	SetHeaderExtra(wantUncle, &HeaderExtra{})
+	wantBody := &Body{
+		Transactions: []*Transaction{tx},
+		Uncles:       []*Header{wantUncle},
+	}
+	wantBodyExtra := &BlockBodyExtra{
+		Version: 8,
+		ExtData: ptrTo([]byte{9}),
+	}
+	extras.Body.Set(wantBody, wantBodyExtra)
+
+	body := block.Body()
+	assert.Equal(t, wantBody, body)
+	bodyExtra := extras.Body.Get(body)
+	assert.Equal(t, wantBodyExtra, bodyExtra)
+
+	exportedFieldsPointToDifferentMemory(t, blockExtras, bodyExtra)
+}
+
+func TestBlockGetters(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		name               string
+		headerExtra        *HeaderExtra
+		blockExtra         *BlockBodyExtra
+		wantExtDataGasUsed *big.Int
+		wantBlockGasCost   *big.Int
+		wantVersion        uint32
+		wantExtData        []byte
+	}{
+		{
+			name:        "empty",
+			headerExtra: &HeaderExtra{},
+			blockExtra:  &BlockBodyExtra{},
+		},
+		{
+			name: "fields_set",
+			headerExtra: &HeaderExtra{
+				ExtDataGasUsed: big.NewInt(1),
+				BlockGasCost:   big.NewInt(2),
+			},
+			blockExtra: &BlockBodyExtra{
+				Version: 3,
+				ExtData: ptrTo([]byte{4}),
+			},
+			wantExtDataGasUsed: big.NewInt(1),
+			wantBlockGasCost:   big.NewInt(2),
+			wantVersion:        3,
+			wantExtData:        []byte{4},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			header := &Header{}
+			SetHeaderExtra(header, test.headerExtra)
+
+			block := NewBlock(header, nil, nil, nil, stubHasher{})
+			extras.Block.Set(block, test.blockExtra)
+
+			extData := BlockExtData(block)
+			assert.Equal(t, test.wantExtData, extData)
+
+			version := BlockVersion(block)
+			assert.Equal(t, test.wantVersion, version)
+
+			extDataGasUsed := BlockExtDataGasUsed(block)
+			assert.Equal(t, test.wantExtDataGasUsed, extDataGasUsed)
+
+			blockGasCost := BlockGasCost(block)
+			assert.Equal(t, test.wantBlockGasCost, blockGasCost)
+		})
+	}
+}
+
+func TestNewBlockWithExtData(t *testing.T) {
+	t.Parallel()
+
+	// This transaction is generated beforehand because of its unexported time field being set
+	// on creation.
+	testTx := NewTransaction(0, common.Address{1}, big.NewInt(2), 3, big.NewInt(4), []byte{5})
```

Comment:
(optional) If you use `cmp.Diff()` then you can fine-tune your comparisons instead of the one-size-fits-all `assert.Equal()` and `reflect.DeepEqual()`.

Reply:
Tests refactored and this is no longer needed, resolving this!